### PR TITLE
tidb-in-kubernetes: add maintain-node documentation

### PR DESCRIPTION
**Pull request by aylei**
Preview: https://github.com/aylei/docs-cn/blob/maintain-node/tidb-on-kubernetes/maintain-node.md

@tennix @cofyc @lilin90 @dcalvin PTAL

Signed-off-by: Aylei <rayingecho@gmail.com>

### CODE REVIEW

File: tidb-on-kubernetes/maintain-node.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
+---
```
Comment (lilin90) [inline tidb-on-kubernetes/maintain-node.md]: In https://github.com/pingcap/docs-cn/pull/1419, the directory name is `tidb-in-kubernetes`. Please keep it consistent. Use `tidb-in-kubernetes`.

File: tidb-on-kubernetes/maintain-node.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+    ```shell
+    $ watch pd-ctl -d store ${ID}
+    ```
+
+5. 解除 TiKV 实例与节点本地盘的绑定：
+
+    ```shell
+    # 查询 Pod 使用的 PesistentVolumeClaim
+    $ kubectl get -n ${namespace} pod ${pod_name} -ojson | jq '.spec.volumes | .[] | select (.name == "tikv") | .persistentVolumeClaim.claimName'
```
Comment (cofyc) [inline tidb-on-kubernetes/maintain-node.md]: `jq` is not installed in Linux distribution by default. We need to ask the user to install `jq` tool first. For this command, how about using kubectl jsonpath? Here is an example:

```
# kubectl -n stability-cluster1 get pods stability-cluster1-tikv-0 -ojsonpath='{.spec.volumes[?(@.name=="tikv")].persistentVolumeClaim.claimName}{"\n"}'
tikv-stability-cluster1-tikv-0
```

File: tidb-on-kubernetes/maintain-node.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    
+    # 删除该 PesistentVolumeClaim
+    $ kubectl delete pvc ${pvc_name}
```
Comment (cofyc) [inline tidb-on-kubernetes/maintain-node.md]: `-n ${namespace}`

File: tidb-on-kubernetes/maintain-node.md (hunk 2 above)
Comment (cofyc) [reply tidb-on-kubernetes/maintain-node.md]: btw, I'm fine with `jq` version too. `jq` is more advanced than kubectl `-ojsonpath` and does more tasks. Maybe we can add a link to jq installation guide somewhere?

File: tidb-on-kubernetes/maintain-node.md (hunk 1 above)
Comment (aylei) [reply tidb-on-kubernetes/maintain-node.md]: good catch

File: tidb-on-kubernetes/maintain-node.md (hunk 2 above)
Comment (aylei) [reply tidb-on-kubernetes/maintain-node.md]: Good catch, I prefer to guide user to install `jq`, which is more ergonomic

File: tidb-in-kubernetes/maintain-node.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
+---
+title: 维护节点
+category: how-to
+---
+
+# 维护节点
+
+TiDB 是高可用数据库，可以在部分数据库节点下线的情况下正常运行，因此，我们可以安全地对底层 Kubernetes 节点进行停机维护。在具体操作时，针对 PD/TiKV/TiDB 实例的不同特性，我们还需要采取不同的操作策略。本文档将详细介绍如何对 Kuberentes 节点进行临时或长期的维护操作。
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
TiDB 是高可用数据库，可以在部分数据库节点下线的情况下正常运行，因此，我们可以安全地对底层 Kubernetes 节点进行停机维护。在具体操作时，针对 PD、TiKV、TiDB 实例的不同特性，我们还需要采取不同的操作策略。本文档将详细介绍如何对 Kuberentes 节点进行临时或长期的维护操作。
```

File: tidb-in-kubernetes/maintain-node.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+环境准备：除了常规的 `kubectl`, `tkctl` 外，下面的操作还会用到 `jq`，可以参考 [Download jq](https://stedolan.github.io/jq/download/) 进行安装。
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
环境准备：

- `kubectl`
- `tkctl`
- [`jq`](https://stedolan.github.io/jq/download/)
```
建议改成 list 来列出需要安装的工具，更方便用户查看。另外，请补出`kubectl`和`tkctl`的安装链接。

File: tidb-in-kubernetes/maintain-node.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+> **注意：**
+>
+> 长期维护节点前，需要保证 Kubernetes 集群的剩余资源足够运行 TiDB 集群。
+
+## 维护 PD/TiDB 实例所在节点
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
## 维护 PD 和 TiDB 实例所在节点
```

File: tidb-in-kubernetes/maintain-node.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+PD/TiDB 实例的迁移较快，可以采取主动驱逐实例到其它节点上的策略进行节点维护：
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
PD 和 TiDB 实例的迁移较快，可以采取主动驱逐实例到其它节点上的策略进行节点维护：
```

File: tidb-in-kubernetes/maintain-node.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+1. 检查待维护节点上是否 TiKV 实例，假如存在 TiKV 实例，请参考 [维护 TiKV 实例所在节点](#维护-tikv-实例所在节点):
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
1. 检查待维护节点上是否有 TiKV 实例，假如存在 TiKV 实例，请参考[维护 TiKV 实例所在节点](#维护-tikv-实例所在节点):
```

File: tidb-in-kubernetes/maintain-node.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    $ kubectl cordon <node-name>
+    ```
+
+3. 使用 `kubectl drain` 命令将待维护节点上的数据库实例迁移到其它实例上：
+
+    ```shell
+    $ kubectl drain <node-name> --ignore-daemonsets --delete-local-data
+    ```
+
+    运行后，该节点上的 TiDB 实例会自动迁移到其它可用节点上，PD 实例则会在 5 分钟后触发[自动故障转移](/TODO/link-to-auto-failover)补齐节点。
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]: This link is not working. Please check it again.

File: tidb-in-kubernetes/maintain-node.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    
+    观察到节点进入 `Ready` 状态后，继续操作。
+
+5. 使用 `kubectl uncordon` 命令解除节点的调度限制:
+
+    ```shell
+    $ kubectl uncordon <node-name>
+    ```
+
+6. 观察 Pod 是否全部恢复正常运行
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
6. 观察 Pod 是否全部恢复正常运行：
```

File: tidb-in-kubernetes/maintain-node.md [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+
+    ```shell
+    $ watch kubectl get -n $namespace pod -o wide
+    # 或者
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
    # 或者：
```

File: tidb-in-kubernetes/maintain-node.md [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+### 短期内可恢复的节点维护
+
+针对短期维护，我们可以通过调整 PD 集群的 `max-store-down-time` 配置来增大集群所允许的 TiKV 实例下线时间，在此时间内维护完毕恢复 Kubernetes 节点后，所有该节点上的 TiKV 实例会自动恢复。
+
+    ```shell
+    $ kubectl port-forward svc/${CLUSTER_NAME}-pd 2379:2379
+    $ pd-ctl -d 
+    ```
+
+调整 `max-store-donw-time` 到合理的的值后，后续的操作方式与 [维护 PD/TiDB 实例所在节点](#维护-pdtidb-实例所在节点) 相同。
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
调整 `max-store-donw-time` 到合理的值后，后续的操作方式与[维护 PD 和 TiDB 实例所在节点](#维护-pd-和-tidb-实例所在节点) 相同。 
```

File: tidb-in-kubernetes/maintain-node.md [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+1. 使用 `kubectl cordon` 命令防止新的 Pod 调度到待维护节点上：
+
+    ```shell
+    $ kubectl cordon <node-name>
+    ```
+
+2. 查看待维护节点上的 TiKV 实例:
+
+    ```shell
+	$ tkctl get -A tikv | grep <node-name>
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
    $ tkctl get -A tikv | grep <node-name>
```

File: tidb-in-kubernetes/maintain-node.md [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+    ```
+
+3. 使用 `pd-ctl` 主动下线 TiKV 实例 
+
+    > **注意：**
+    > 
+    > 下线 TiKV 实例前，需要保证集群中剩余的 TiKV 实例数不少于 PD 配置中的 TiKV 数据副本数（配置项：max-replicas）。假如不符合该条件，需要先操作[扩容 TiKV](/TODO/link-to-scaling)。
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
    > 下线 TiKV 实例前，需要保证集群中剩余的 TiKV 实例数不少于 PD 配置中的 TiKV 数据副本数（配置项：`max-replicas`）。假如不符合该条件，需要先操作[扩容 TiKV](/TODO/link-to-scaling)。
```
Make sure this link is working before merging this PR.

File: tidb-in-kubernetes/maintain-node.md [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+
+    ```shell
+    # 查看 TiKV 实例的 store-id
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
    # 查看 TiKV 实例的 `store-id`。
```
A period is needed at the end of a code comment sentence.

File: tidb-in-kubernetes/maintain-node.md [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+    $ kubectl get tc ${CLUSTER_NAME} -ojson | jq '.status.tikv.stores | .[] | select ( .podName == "${POD_NAME}" ) | .id'
+
+    # 下线实例
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
    # 下线实例。
```

File: tidb-in-kubernetes/maintain-node.md [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+    $ kubectl port-forward svc/${CLUSTER_NAME}-pd 2379:2379
+    $ pd-ctl -d store delete ${ID}
+    ```
+
+4. 等待 store 状态(`state_name`)转移为 `Tombstone`：
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
4. 等待 store 状态（`state_name`）转移为 `Tombstone`：
```

File: tidb-in-kubernetes/maintain-node.md [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+
+    ```shell
+    $ watch pd-ctl -d store ${ID}
+    ```
+
+5. 解除 TiKV 实例与节点本地盘的绑定：
+
+    ```shell
+    # 查询 Pod 使用的 PesistentVolumeClaim
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
    # 查询 Pod 使用的 `PesistentVolumeClaim`。
```

File: tidb-in-kubernetes/maintain-node.md [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+    $ kubectl get -n ${namespace} pod ${pod_name} -ojson | jq '.spec.volumes | .[] | select (.name == "tikv") | .persistentVolumeClaim.claimName'
+    
+    # 删除该 PesistentVolumeClaim
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
    # 删除该 `PesistentVolumeClaim`。
```

File: tidb-in-kubernetes/maintain-node.md [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+    $ kubectl delete -n ${namespace} pod ${pod_name}
+    ```
+
+7. 观察该 TiKV 实例是否正常调度到其它节点上：
+
+    ```shell
+    $ watch kubectl -n ${namespace} get pod -o wide
+    ```
+	
+	假如待维护节点上还有其它 TiKV 实例，则重复同样的操作步骤直到所有的 TiKV 实例都迁移到其它节点上。
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
    假如待维护节点上还有其它 TiKV 实例，则重复同样的操作步骤直到所有的 TiKV 实例都迁移到其它节点上。
```

File: tidb-in-kubernetes/maintain-node.md [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+
+8. 确认节点不再有 TiKV 实例后，再逐出节点上的其它实例：
+
+    ```shell
+    $ kubectl drain <node-name> --ignore-daemonsets --delete-local-data
+    ```
+
+9. 再次确认节点不再有任何 TiKV/TiDB/PD 实例运行：
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
9. 再次确认节点不再有任何 TiKV、TiDB 和 PD 实例运行：
```

File: tidb-in-kubernetes/maintain-node.md [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+
+    ```shell
+    $ kubectl get pod --all-namespaces | grep <node-name>
+    ```
+
+10. 最后（可选），假如是长期下线节点，建议将节点从 Kubernetes 集群中删除：
+	
+    ```shell
+	$ kuebctl delete node <node-name>
```
Comment (cocoyi17) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
    $ kuebctl delete node <node-name>
```

File: tidb-in-kubernetes/maintain-node.md [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+    $ watch tkctl get all
+    ```
+
+    Pod 恢复正常运行后，操作结束。
+
+## 维护 TiKV 实例所在节点
+
+TiKV 实例迁移较慢，并且会对集群造成一定的数据迁移负载，因此在维护 TiKV 实例所在节点前，需要根据维护需求选择操作策略：假如是短期节点维护，则不迁移 TiKV 节点，在维护结束后原地恢复节点；假如是短期内不可恢复的节点维护，则需要规划 TiKV 的迁移工作。
+
+### 短期内可恢复的节点维护
```
Comment (yikeke) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
### 维护短期内可恢复的节点
```
Other headings are "verb + noun" phrases, please try to keep consistent with them to improve readability.

File: tidb-in-kubernetes/maintain-node.md [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+针对短期维护，我们可以通过调整 PD 集群的 `max-store-down-time` 配置来增大集群所允许的 TiKV 实例下线时间，在此时间内维护完毕恢复 Kubernetes 节点后，所有该节点上的 TiKV 实例会自动恢复。
+
+    ```shell
+    $ kubectl port-forward svc/${CLUSTER_NAME}-pd 2379:2379
+    $ pd-ctl -d 
+    ```
+
+调整 `max-store-donw-time` 到合理的的值后，后续的操作方式与 [维护 PD/TiDB 实例所在节点](#维护-pdtidb-实例所在节点) 相同。 
+
+### 短期内不可恢复的节点维护
```
Comment (yikeke) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
### 维护短期内不可恢复的节点
```
ditto

File: tidb-in-kubernetes/maintain-node.md (hunk 9 above)
Comment (aylei) [reply tidb-in-kubernetes/maintain-node.md]: This is intended, the target document has not completed yet

File: tidb-in-kubernetes/maintain-node.md (hunk 14 above)
Comment (aylei) [reply tidb-in-kubernetes/maintain-node.md]: This link is a placeholder (the scaling documentation has not completed yet)

File: tidb-in-kubernetes/maintain-node.md (hunk 9 above)
Comment (yikeke) [reply tidb-in-kubernetes/maintain-node.md]: OK.

File: tidb-in-kubernetes/maintain-node.md [hunk 25]
```diff
@@ -0,0 +1,169 @@
...
+### 维护短期内可恢复的节点
+
+针对短期维护，我们可以通过调整 PD 集群的 `max-store-down-time` 配置来增大集群所允许的 TiKV 实例下线时间，在此时间内维护完毕恢复 Kubernetes 节点后，所有该节点上的 TiKV 实例会自动恢复。
+
+    ```shell
+    $ kubectl port-forward svc/${CLUSTER_NAME}-pd 2379:2379
+    $ pd-ctl -d 
+    ```
+
+调整 `max-store-donw-time` 到合理的值后，后续的操作方式与[维护 PD 和 TiDB 实例所在节点](#维护-pd-和-tidb-实例所在节点) 相同。
```
Comment (yikeke) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
调整 `max-store-donw-time` 到合理的值后，后续的操作方式与[维护 PD 和 TiDB 实例所在节点](#维护-pd-和-tidb-实例所在节点)相同。 
```

File: tidb-in-kubernetes/maintain-node.md [hunk 26]
```diff
@@ -0,0 +1,169 @@
...
+
+针对短期内不可恢复的节点维护，如节点长期下线等情形，需要使用 `pd-ctl` 主动通知 TiDB 集群下线对应的 TiKV 实例，再手动解除 TiKV 实例与该节点的绑定。
+
+1. 使用 `kubectl cordon` 命令防止新的 Pod 调度到待维护节点上：
+
+    ```shell
+    $ kubectl cordon <node-name>
+    ```
+
+2. 查看待维护节点上的 TiKV 实例:
```
Comment (yikeke) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
2. 查看待维护节点上的 TiKV 实例：
```

File: tidb-in-kubernetes/maintain-node.md [hunk 27]
```diff
@@ -0,0 +1,169 @@
...
+
+    ```shell
+    $ tkctl get -A tikv | grep <node-name>
+    ```
+
+3. 使用 `pd-ctl` 主动下线 TiKV 实例
```
Comment (yikeke) [inline tidb-in-kubernetes/maintain-node.md]:
```suggestion
3. 使用 `pd-ctl` 主动下线 TiKV 实例。
```

File: tidb-in-kubernetes/maintain-node.md [hunk 28]
```diff
@@ -0,0 +1,169 @@
...
+
+    ```shell
+    $ tkctl get -A tikv | grep <node-name>
+    ```
+
+3. 使用 `pd-ctl` 主动下线 TiKV 实例 
+
+    > **注意：**
+    > 
+    > 下线 TiKV 实例前，需要保证集群中剩余的 TiKV 实例数不少于 PD 配置中的 TiKV 数据副本数（配置项：`max-replicas`）。假如不符合该条件，需要先操作[扩容 TiKV](/TODO/link-to-scaling)。
```
Comment (yikeke) [inline tidb-in-kubernetes/maintain-node.md]: This link is a dead link too. Please remove it before you merge this PR. You can add it back when /TODO/link-to-scaling is ready.

**Comment by yikeke** [comment]
I have updated all code blocks in this file to apply the new code block format rules, see details in [Update code block format to better support copy](https://internal.pingcap.net/jira/browse/ICM-95). 
PTAL @aylei

File: tidb-in-kubernetes/maintain-node.md [hunk 29]
```diff
@@ -0,0 +1,233 @@
+---
```
Comment (yikeke) [inline tidb-in-kubernetes/maintain-node.md]: Please move this file to docs-cn/how-to/maintain/kubernetes-operations/ directory. Thanks!

**Comment by aylei** [comment]
> I have updated all code blocks in this file to apply the new code block format rules, see details in [Update code block format to better support copy](https://internal.pingcap.net/jira/browse/ICM-95).
> PTAL @aylei

Awesome, I will update this PR accordingly

**Comment by yikeke** [comment]
> > I have updated all code blocks in this file to apply the new code block format rules, see details in [Update code block format to better support copy](https://internal.pingcap.net/jira/browse/ICM-95).
> > PTAL @aylei
> 
> Awesome, I will update this PR accordingly

I have updated this PR before, please address other comments in this PR. We should merge this PR today. Thanks!

**Comment by aylei** [comment]
@yikeke I've addressed all the comments, thanks for your patient review

**Pull request merged**